go-fuzz-dep/main.go

```diff
@@ -21,97 +17,24 @@ import (
 type Bool = bool
 
 var (
-	inFD  FD
-	outFD FD
-
 	CoverTab    *[CoverSize]byte
-	input       []byte
 	sonarRegion []byte
 	sonarPos    uint32
+	CoverTabTmp [CoverSize]byte
 )
 
-func init() {
-	var mem []byte
-	mem, inFD, outFD = setupCommFile()
-	CoverTab = (*[CoverSize]byte)(unsafe.Pointer(&mem[0]))
-	input = mem[CoverSize : CoverSize+MaxInputSize]
-	sonarRegion = mem[CoverSize+MaxInputSize:]
-}
 
-func Main(f func([]byte) int) {
-	runtime.GOMAXPROCS(1) // makes coverage more deterministic, we parallelize on higher level
-	for {
-		n := read(inFD)
-		if n > uint64(len(input)) {
-			println("invalid input length")
-			syscall.Exit(1)
-		}
-		for i := range CoverTab {
-			CoverTab[i] = 0
-		}
-		atomic.StoreUint32(&sonarPos, 0)
-		t0 := time.Now()
-		res := f(input[:n])
-		ns := time.Since(t0)
-		write(outFD, uint64(res), uint64(ns), uint64(atomic.LoadUint32(&sonarPos)))
-	}
-}
 
-// read reads little-endian-encoded uint64 from fd.
-func read(fd FD) uint64 {
-	rd := 0
-	var buf [8]byte
-	for rd != len(buf) {
-		n, err := fd.read(buf[rd:])
-		if err == syscall.EINTR {
-			continue
-		}
-		if n == 0 {
-			syscall.Exit(1)
-		}
-		if err != nil {
-			println("failed to read fd =", fd, "errno =", err.(syscall.Errno))
-			syscall.Exit(1)
-		}
-		rd += n
-	}
-	return deserialize64(buf[:])
+func init() {
+	CoverTab = (*[CoverSize]byte)(unsafe.Pointer(&CoverTabTmp[0]))
```

Comment:
It would be super nice to put this directly into go-fuzz-dep package:
```
// #cgo CFLAGS: -Wall -Werror
/*
#ifdef __linux__
__attribute__((weak, section("__libfuzzer_extra_counters")))
#else
#error Currently only Linux is supported
#endif
unsigned char __go_fuzz_counters[65536];
*/
import "C"

func init() {
	CoverTab = (*[CoverSize]byte)(unsafe.Pointer(&C.__go_fuzz_counters[0]))
}
```

Then would not need bootstrap CoverTab and the Initialize function.
But unfortunately cgo can't be used here:

```
go-fuzz$ go install ./... && go-fuzz-build -libfuzzer ../go-fuzz-corpus/fmt && clang fmt.a -fsanitize=fuzzer
gopath/src/github.com/dvyukov/go-fuzz/go-fuzz-dep/main.go:9:2: could not import C (no metadata for C)
gopath/src/github.com/dvyukov/go-fuzz/go-fuzz-dep/main.go:24:8: C redeclared in this block
gopath/src/github.com/dvyukov/go-fuzz/go-fuzz-dep/main.go:9:2: 	other declaration of C
typechecking of ../go-fuzz-corpus/fmt failed
```

@josharian do you know if it's theoretically possible to support cgo here?
It's not a super big deal because the current scheme works. But if it's easy to do, then it would be nice.